2021_RTCSA_dynfed/src/dynfed.rs

```diff
@@ -48,6 +122,108 @@ where
 
     (minimum_cores, execution_order)
 }
+
+#[allow(dead_code)]
+pub fn dynamic_federated<T>(
+    dag_set: &mut [Graph<NodeData, i32>],
+    processor: &mut T,
+    scheduler: &mut impl SchedulerBase<T>,
+) -> i32
+where
+    T: ProcessorBase + Clone,
+{
+    let mut current_time = 0;
+    let dag_set_length = dag_set.len();
+    let mut dyn_fed_handlers = vec![DynamicFederatedHandler::new(); dag_set_length];
+    let mut dag_queue: VecDeque<Graph<NodeData, i32>> = VecDeque::new();
+    let mut finished_dags_count = 0;
+
+    scheduler.set_processor(processor);
+
+    for (dag_id, dag) in dag_set.iter_mut().enumerate() {
+        dag.set_dag_id(dag_id);
+        dyn_fed_handlers[dag_id].set_minimum_cores_and_execution_order(dag, scheduler);
+    }
+
+    while finished_dags_count < dag_set_length {
+        dag_set
+            .iter_mut()
+            .filter(|dag| current_time == dag.get_head_offset())
+            .for_each(|dag| dag_queue.push_back(dag.clone()));
+
+        for dynamic_federated_handler in dyn_fed_handlers.iter_mut() {
+            dynamic_federated_handler.update_using_cores();
+        }
+
+        //Start DAG if there are enough free core
+        while let Some(dag) = dag_queue.front() {
+            let dag_id = dag.get_dag_id();
+            let processor_cores = processor.get_number_of_cores() as i32;
+            let allocated_cores = get_total_allocated_cores(&dyn_fed_handlers);
+
+            if dyn_fed_handlers[dag_id].can_start(processor_cores, allocated_cores) {
+                dag_queue.pop_front();
+                dyn_fed_handlers[dag_id].start();
+            } else {
+                break;
+            }
+        }
+
+        //Assign a node per DAG
+        for (dag_id, dag) in dag_set.iter_mut().enumerate() {
```

Comment:
The reason for this change is that we thought this would be better performance than making a function call every time.

However, since it did not seem to affect performance that much, it was implemented as suggested.

Reply:
The previous implementation only works if the index in `dag_set` and dag_id are the same.
Such unstable implementations should be avoided.
It is also confusing to reviewers because it is not obvious from looking at this section alone that index and dag_id are the same.